src/Forms/HTMLEditor/HTMLEditorRuleSet.php

```diff
@@ -0,0 +1,173 @@
+<?php
+
+namespace SilverStripe\Forms\HTMLEditor;
+
+use DOMElement;
+use InvalidArgumentException;
+
+/**
+ * A set of rules to determine which elements and attributes are allowed for a given HTMLEditorConfig.
+ * This ruleset will be used by HTMLEditorSanitiser for server-side sanitisation of HTML.
+ */
+class HTMLEditorRuleSet
+{
+    private HTMLEditorElementRule $globalRule;
+
+    /**
+     * Associative array of element rules which are included in this ruleset.
+     * Excludes rules which use regex - see $elementPatternRules
+     * @var HTMLEditorElementRule[]
+     */
+    private array $elementRules = [];
+
+    /**
+     * Associative array of element rules with regex expressions which are included in this ruleset.
+     * See also $elementRules
+     * @var HTMLEditorElementRule[]
+     */
+    private array $elementPatternRules = [];
+
+    private array $elementSubstitutionRules = [];
+
+    /**
+     * Check if the name given for an element or attribute rule is a pattern.
+     * If so, it should be passed through patternToRegex() before being assigned to the rule.
+     */
+    public static function nameIsPattern(string $name): bool
+    {
+        $hasPatternsRegExp = '/[*?+]/';
+        return preg_match($hasPatternsRegExp, $name);
+    }
+
+    /**
+     * Given a TinyMCE pattern (close to unix glob style), create a regex that does the match
+     *
+     * @param $pattern - The TinyMCE pattern
+     * @return string - The equivalent regex
+     */
+    public static function patternToRegex(string $pattern): string
+    {
+        return '/^' . preg_replace('/([?+*])/', '.$1', $pattern) . '$/';
+    }
+
+    /**
+     * Undocumented function
+     *
+     * @param string $regex
+     * @return string
```

Comment:
Probably delete the docblock param / return type

Reply:
Not sure what method this was but I've documented all methods with "Undocumented function"